### PR TITLE
feat: FW-1060 post translation

### PR DESCRIPTION
**Pull request by dudu0506**
https://mask.atlassian.net/browse/FW-1060

# Pull Request

## Related Issue

Closes FW-0000

## Checklist

-   [ ] I have tested these changes locally.
-   [ ] I have reviewed the code changes.
-   [ ] I have added unit tests if applicable.

### CODE REVIEW

File: src/helpers/getBrowserLanguage.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  const browserLanguage = getBrowserLanguage();
+  const browserLanguageLowerCase = browserLanguage.toLowerCase();
+  const localLowerCase = local.toLowerCase();
+  if (browserLanguageLowerCase === localLowerCase) {
+    return true;
+  }
+  if (browserLanguageLowerCase.startsWith(localLowerCase)) {
+    return true;
+  }
+  return false;
```
Comment (nuanyang233) [inline src/helpers/getBrowserLanguage.ts:17]:
```suggestion
return browserLanguageLowerCase.startsWith(localLowerCase)
```

File: src/helpers/getLangNameFromLocal.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { Language } from '@/constants/enum.js';
+
+export const getLangNameFromLocal = (local: string, failedValue = local) => {
+  const matchedLang = Object.entries(Language).find(([_, value]) => value === local);
+  return matchedLang?.[0] ?? failedValue;
```
Comment (nuanyang233) [inline src/helpers/getLangNameFromLocal.ts:5]: Plz use `first` from `lodash-es`

File: src/helpers/getBrowserLanguage.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import type { Language } from "@/constants/enum.js";
+
+export const getBrowserLanguage = () => {
+  return navigator.language as unknown as Language;
```
Comment (guanbinrui) [inline src/helpers/getBrowserLanguage.ts:4]: It's [RFC5646](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/language).

File: src/helpers/getBrowserLanguage.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import type { Language } from "@/constants/enum.js";
+
+export const getBrowserLanguage = () => {
+  return navigator.language as unknown as Language;
+}
+
+export const isSameLanguageWithBrowser = (local: string) => {
```
Comment (guanbinrui) [inline src/helpers/getBrowserLanguage.ts:7]:
```suggestion
export const isSameLanguageWithBrowser = (locale: string) => {

```

File: src/components/Posts/ContentTranslator.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    const isValidContentLang = !!contentLanguage && contentLanguage !== 'N/A';
+    const isSameLanguage = isValidContentLang && isSameLanguageWithBrowser(contentLanguage);
+
+    if (!isLogin || !isValidContentLang || isSameLanguage) return null;
+
+    return (
+        <>
+            <div className='my-1.5'>
+                <ClickableButton className="text-sm text-[#8E96FF]" onClick={onTranslate}>
```
Comment (guanbinrui) [inline src/components/Posts/ContentTranslator.tsx:69]: Try to pick a color value from palette.

File: src/components/Posts/ContentTranslator.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            const result = await translate(browserLanguage, content);
+            return result;
+        },
+    })
+
+    const buttonLabel = useMemo(() => {
+        const result = data?.translations?.[0]?.text;
+        if (result && enableTranslate) {
+            const detectedLanguage = data?.detectedLanguage;
+            const append = detectedLanguage ? ` from ${getLangNameFromLocal(detectedLanguage)}` : ''
```
Comment (guanbinrui) [inline src/components/Posts/ContentTranslator.tsx:41]: i18n

File: src/components/Posts/ContentTranslator.tsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    })
+    const { data, isLoading, refetch } = useQuery({
+        queryKey: ['content-translator', cacheKey],
+        enabled: enableTranslate,
+        queryFn: async () => {
+            const browserLanguage = getBrowserLanguage();
+            const result = await translate(browserLanguage, content);
+            return result;
+        },
+    })
```
Comment (guanbinrui) [inline src/components/Posts/ContentTranslator.tsx:35]: Merge those hooks.

File: src/components/Posts/ContentTranslator.tsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    const buttonLabel = useMemo(() => {
```
Comment (guanbinrui) [inline src/components/Posts/ContentTranslator.tsx:37]: Remove memo.

File: src/components/Posts/ContentTranslator.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        const result = data?.translations?.[0]?.text;
+        if (result && enableTranslate) {
+            const detectedLanguage = data?.detectedLanguage;
+            const append = detectedLanguage ? ` from ${getLangNameFromLocal(detectedLanguage)}` : ''
+            return t`Translated${append}`;
+        }
+        return isLoading ? t`Translating...` : t`Translate post`;
+    }, [isLoading, data, enableTranslate])
+    const translatedContent = useMemo(() => {
```
Comment (guanbinrui) [inline src/components/Posts/ContentTranslator.tsx:46]: Remove meme.

File: src/components/Posts/ContentTranslator.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                return { contentLanguage: null, translated: null };
+            }
+        },
+        [content]
+    )
+
+    useEffect(() => {
+        if (!isLogin) return;
+        handleTranslate(true);
+    }, [isLogin])
```
Comment (guanbinrui) [inline src/components/Posts/ContentTranslator.tsx:47]:
```suggestion
```

File: src/components/Posts/ContentTranslator.tsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+    const translatedText = data?.translated?.translations?.[0]?.text;
+
+    const buttonLabel = (() => {
+        if (translatedText) {
+            const detectedLanguage = data?.translated?.detectedLanguage;
+            const contentLangName = getLangNameFromLocal(detectedLanguage ?? '');
+            return contentLangName ? t`Translated from ${contentLangName}` : t`Translated`;
+        }
+        return loading ? t`Translating...` : t`Translate post`;
```
Comment (guanbinrui) [inline src/components/Posts/ContentTranslator.tsx:57]: Please handle error: Failed to translate.

File: src/components/Posts/PostBody.tsx [hunk 12]
```diff
@@ -67,6 +70,22 @@ export const PostBody = forwardRef<HTMLDivElement, PostBodyProps>(function PostB
 
     const muted = useIsMuted(post.author);
 
+    const postContent = (endingLinkCollapsed
+        ? removeUrlAtEnd(post.metadata.content?.oembedUrl, post.metadata.content?.content)
+        : post.metadata.content?.content) ?? '';
+
+    const renderPostContent = useCallback((content: string) => (
+        <Markup
```
Comment (guanbinrui) [inline src/components/Posts/PostBody.tsx:78]: Create a component: `<PostMarkup />`, so we can reduce a function property.

**Pull request merged**